### PR TITLE
Use concrete RestResponse in generator

### LINKED ISSUE #153: Generating RestResponse subclasses
`Single<RestResponse<MyGetHeaders, Flowable<ByteBuffer>>> getWithRestResponseAsync()` has quite a daunting return type.
I think we should consider generating simply `Single<MyGetResponse> getAsync()` where MyGetResponse is:

```java
public final class MyGetResponse extends RestResponse<MyGetHeaders, Flowable<ByteBuffer>> {
    MyGetResponse(int statusCode, MyGetHeaders headers, Map<String, String> rawHeaders, Flowable<ByteBuffer> body) {
        super(...);
    }
}
```
It might be good to explicitly implement the members in the child class because when people go to definition on one of the inherited members and they only see TGeneric instead of the concrete type being used in context, it's going to be annoying and hard to follow what the types of things are in the code.

This should be done before we release a preview.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #153. Will need to merge Azure/autorest-clientruntime-for-java#393 and publish a snapshot before CI will pass.

### CODE REVIEW

**Comment by RikkiGibson** [comment]
> linux-x64 restart

**Comment by RikkiGibson** [comment]
> win32-x64 restart

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.14*

## ❌️ General: 92%

<details><summary>38 out of 522 features are not covered by tests</summary><p>

[`ConstantsInBody`](https://github.com/Azure/autorest.testserver/search?q=ConstantsInBody), [`ConstantsInPath`](https://github.com/Azure/autorest.testserver/search?q=ConstantsInPath), [`FileStreamVeryLarge`](https://github.com/Azure/autorest.testserver/search?q=FileStreamVeryLarge), [`FormdataStreamUploadFile`](https://github.com/Azure/autorest.testserver/search?q=FormdataStreamUploadFile), [`HttpRedirect300Get`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect300Get), [`HttpRedirect300Head`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect300Head), [`HttpRedirect301Get`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect301Get), [`HttpRedirect301Put`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect301Put), [`HttpRedirect302Get`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect302Get), [`HttpRedirect302Head`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect302Head), [`HttpRedirect302Patch`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect302Patch), [`HttpRedirect303Post`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect303Post), [`HttpRedirect307Delete`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Delete), [`HttpRedirect307Get`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Get), [`HttpRedirect307Head`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Head), [`HttpRedirect307Patch`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Patch), [`HttpRedirect307Post`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Post), [`HttpRedirect307Put`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Put), [`ResponsesScenarioC200MatchingModel`](https://github.com/Azure/autorest.testserver/search?q=ResponsesScenarioC200MatchingModel), [`ResponsesScenarioC201MatchingModel`](https://github.com/Azure/autorest.testserver/search?q=ResponsesScenarioC201MatchingModel), [`ResponsesScenarioC404MatchingModel`](https://github.com/Azure/autorest.testserver/search?q=ResponsesScenarioC404MatchingModel), [`StreamUploadFile`](https://github.com/Azure/autorest.testserver/search?q=StreamUploadFile), [`UrlPathsArrayCSVInPath`](https://github.com/Azure/autorest.testserver/search?q=UrlPathsArrayCSVInPath), [`UrlQueriesArrayMultiEmpty`](https://github.com/Azure/autorest.testserver/search?q=UrlQueriesArrayMultiEmpty), [`UrlQueriesArrayMultiNull`](https://github.com/Azure/autorest.testserver/search?q=UrlQueriesArrayMultiNull), [`UrlQueriesArrayMultiValid`](https://github.com/Azure/autorest.testserver/search?q=UrlQueriesArrayMultiValid), [`allowedValueEnum`](https://github.com/Azure/autorest.testserver/search?q=allowedValueEnum), [`expectedEnum`](https://github.com/Azure/autorest.testserver/search?q=expectedEnum), [`getByteEmpty`](https://github.com/Azure/autorest.testserver/search?q=getByteEmpty), [`getByteInvalid`](https://github.com/Azure/autorest.testserver/search?q=getByteInvalid), [`getComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=getComplexPolymorphismComplicated), [`getEnumReferenced`](https://github.com/Azure/autorest.testserver/search?q=getEnumReferenced), [`getEnumReferencedConstant`](https://github.com/Azure/autorest.testserver/search?q=getEnumReferencedConstant), [`putComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=putComplexPolymorphismComplicated), [`putEnumReferenced`](https://github.com/Azure/autorest.testserver/search?q=putEnumReferenced), [`putEnumReferencedConstant`](https://github.com/Azure/autorest.testserver/search?q=putEnumReferencedConstant), [`roundTripEnum`](https://github.com/Azure/autorest.testserver/search?q=roundTripEnum), [`unexpectedEnum`](https://github.com/Azure/autorest.testserver/search?q=unexpectedEnum)</p></details>

## ❌️ Azure: 94%

<details><summary>7 out of 118 features are not covered by tests</summary><p>

[`AzureMethodQueryUrlEncoding`](https://github.com/Azure/autorest.testserver/search?q=AzureMethodQueryUrlEncoding), [`AzureODataFilter`](https://github.com/Azure/autorest.testserver/search?q=AzureODataFilter), [`AzurePathQueryUrlEncoding`](https://github.com/Azure/autorest.testserver/search?q=AzurePathQueryUrlEncoding), [`AzureRequestClientIdInError`](https://github.com/Azure/autorest.testserver/search?q=AzureRequestClientIdInError), [`AzureSwaggerQueryUrlEncoding`](https://github.com/Azure/autorest.testserver/search?q=AzureSwaggerQueryUrlEncoding), [`AzureXmsRequestClientIdNull`](https://github.com/Azure/autorest.testserver/search?q=AzureXmsRequestClientIdNull), [`LRONonRetryPut201Creating400InvalidJson`](https://github.com/Azure/autorest.testserver/search?q=LRONonRetryPut201Creating400InvalidJson)</p></details>

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.14*

## ❌️ General: 92%

<details><summary>38 out of 522 features are not covered by tests</summary><p>

[`ConstantsInBody`](https://github.com/Azure/autorest.testserver/search?q=ConstantsInBody), [`ConstantsInPath`](https://github.com/Azure/autorest.testserver/search?q=ConstantsInPath), [`FileStreamVeryLarge`](https://github.com/Azure/autorest.testserver/search?q=FileStreamVeryLarge), [`FormdataStreamUploadFile`](https://github.com/Azure/autorest.testserver/search?q=FormdataStreamUploadFile), [`HttpRedirect300Get`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect300Get), [`HttpRedirect300Head`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect300Head), [`HttpRedirect301Get`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect301Get), [`HttpRedirect301Put`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect301Put), [`HttpRedirect302Get`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect302Get), [`HttpRedirect302Head`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect302Head), [`HttpRedirect302Patch`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect302Patch), [`HttpRedirect303Post`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect303Post), [`HttpRedirect307Delete`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Delete), [`HttpRedirect307Get`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Get), [`HttpRedirect307Head`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Head), [`HttpRedirect307Patch`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Patch), [`HttpRedirect307Post`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Post), [`HttpRedirect307Put`](https://github.com/Azure/autorest.testserver/search?q=HttpRedirect307Put), [`ResponsesScenarioC200MatchingModel`](https://github.com/Azure/autorest.testserver/search?q=ResponsesScenarioC200MatchingModel), [`ResponsesScenarioC201MatchingModel`](https://github.com/Azure/autorest.testserver/search?q=ResponsesScenarioC201MatchingModel), [`ResponsesScenarioC404MatchingModel`](https://github.com/Azure/autorest.testserver/search?q=ResponsesScenarioC404MatchingModel), [`StreamUploadFile`](https://github.com/Azure/autorest.testserver/search?q=StreamUploadFile), [`UrlPathsArrayCSVInPath`](https://github.com/Azure/autorest.testserver/search?q=UrlPathsArrayCSVInPath), [`UrlQueriesArrayMultiEmpty`](https://github.com/Azure/autorest.testserver/search?q=UrlQueriesArrayMultiEmpty), [`UrlQueriesArrayMultiNull`](https://github.com/Azure/autorest.testserver/search?q=UrlQueriesArrayMultiNull), [`UrlQueriesArrayMultiValid`](https://github.com/Azure/autorest.testserver/search?q=UrlQueriesArrayMultiValid), [`allowedValueEnum`](https://github.com/Azure/autorest.testserver/search?q=allowedValueEnum), [`expectedEnum`](https://github.com/Azure/autorest.testserver/search?q=expectedEnum), [`getByteEmpty`](https://github.com/Azure/autorest.testserver/search?q=getByteEmpty), [`getByteInvalid`](https://github.com/Azure/autorest.testserver/search?q=getByteInvalid), [`getComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=getComplexPolymorphismComplicated), [`getEnumReferenced`](https://github.com/Azure/autorest.testserver/search?q=getEnumReferenced), [`getEnumReferencedConstant`](https://github.com/Azure/autorest.testserver/search?q=getEnumReferencedConstant), [`putComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=putComplexPolymorphismComplicated), [`putEnumReferenced`](https://github.com/Azure/autorest.testserver/search?q=putEnumReferenced), [`putEnumReferencedConstant`](https://github.com/Azure/autorest.testserver/search?q=putEnumReferencedConstant), [`roundTripEnum`](https://github.com/Azure/autorest.testserver/search?q=roundTripEnum), [`unexpectedEnum`](https://github.com/Azure/autorest.testserver/search?q=unexpectedEnum)</p></details>

## ❌️ Azure: 94%

<details><summary>7 out of 118 features are not covered by tests</summary><p>

[`AzureMethodQueryUrlEncoding`](https://github.com/Azure/autorest.testserver/search?q=AzureMethodQueryUrlEncoding), [`AzureODataFilter`](https://github.com/Azure/autorest.testserver/search?q=AzureODataFilter), [`AzurePathQueryUrlEncoding`](https://github.com/Azure/autorest.testserver/search?q=AzurePathQueryUrlEncoding), [`AzureRequestClientIdInError`](https://github.com/Azure/autorest.testserver/search?q=AzureRequestClientIdInError), [`AzureSwaggerQueryUrlEncoding`](https://github.com/Azure/autorest.testserver/search?q=AzureSwaggerQueryUrlEncoding), [`AzureXmsRequestClientIdNull`](https://github.com/Azure/autorest.testserver/search?q=AzureXmsRequestClientIdNull), [`LRONonRetryPut201Creating400InvalidJson`](https://github.com/Azure/autorest.testserver/search?q=LRONonRetryPut201Creating400InvalidJson)</p></details>
